### ISSUE TITLE
Recent Breaking Changes: window.web3 Removal and window.ethereum changes

### ISSUE THREAD

**Issue by rekmarks** [MetaMask/metamask-extension#8077]
### The Breaking Changes are _Live_

As of version `9.0.2` of the MetaMask extension (and `1.0.9` of MetaMask Mobile), the breaking changes are live. See [the migration guide](https://docs.metamask.io/guide/provider-migration.html) for more details.

The below information is only useful for historical purposes.

-----

### Summary

We are removing `window.web3` and making some breaking changes to `window.ethereum`. These changes apply to all version of MetaMask (extension and mobile).

We committed to maintaining the affected features until November 16, 2020. That day has now come, and _the changes may ship at any time_. Due to app store review processes, we can't say exactly when, but we will make an announcement here once we know for sure.

### New Provider API

We shipped a new, *non-breaking* Provider (`window.ethereum`) API in MetaMask 8.0.4 on July 14, 2020, which happened to be MetaMask's 4th birthday 🎊 🦊 🎊. This API is also available through MetaMask mobile.

Please consult [our documentation](https://docs.metamask.io/guide/ethereum-provider.html) to learn how to use the API.

### Upcoming Breaking Changes

These changes will ship on or after **November 16, 2020**. The version number of MetaMask that includes the breaking changes will be announced by this date.

#### `window.web3` Removal

We are going to remove our injected `window.web3`. See #7163 for why, and our [migration guide](https://docs.metamask.io/guide/provider-migration.html#replacing-window-web3) for replacement strategies.

See #9156 for the latest build without `window.web3`, automatically posted by [@metamaskbot](https://github.com/metamaskbot).

#### `window.ethereum` API Changes

We are also going to make _some_ breaking changes to our provider. These include:

- Stop 0-padding `eth_chainId` return values
- Stop emitting `chainIdChanged`, and instead emit `chainChanged`
- Remove `ethereum.publicConfigStore`
- Remove the following experimental methods:
  - `ethereum._metamask.isEnabled`
  - `ethereum._metamask.isApproved`

#### What to Do Next

Please see [the relevant section of our documentation](https://docs.metamask.io/guide/ethereum-provider.html#upcoming-breaking-changes) for details. _You can migrate to protect yourself from these changes today._ Please see our [migration guide](https://docs.metamask.io/guide/provider-migration.html) for details. There's no reason to wait, [migrate today!](https://docs.metamask.io/guide/provider-migration.html)

**Comment by imnisen** [comment]
Hi, does the current version 7.7.7 support the new api? 

When I try to use `ethereum.send('eth_requestAccounts')` replace `ethereum.enable()`, metamask returns `{code: -32601, message: "The method undefined does not exist/is not available"}`.

**Comment by rekmarks** [comment]
Hi @imnisen, the new API is not live; no production version of MetaMask supports it. It will say in this issue when it is.

**Comment by bhayward93** [comment]
Any recommended ways to test that my site will not be affected by this as a Dapp developer?

**Comment by rekmarks** [comment]
@bhayward93, if you subscribe to this issue and [follow us on Twitter](https://twitter.com/metamask_io) you'll be notified when we have a build with only the new API ready. We'll provide instructions for how to test it, and you'll have a full 6 weeks to do it.

**Comment by dmvt** [comment]
> Hi @imnisen, the new API is not live; no production version of MetaMask supports it. It will say in this issue when it is.

Hey @rekmarks @danfinlay can we get a status update? It looks as though this may be live or partially live. Also, I read in the docs that the plan is to be compliant with EIP 1193, however, EIP 1193 [deprecates the `send` method entirely, preferring `request`](https://eips.ethereum.org/EIPS/eip-1193#request). The docs do not show a `request` method in the new api. Will it be supported?

Thanks!

![Screenshot from 2020-04-05 22-29-33](https://user-images.githubusercontent.com/1116695/78508229-cb1d2380-778d-11ea-8e19-9edfdec8367d.png)

**Comment by rekmarks** [comment]
Hi @dmvt, thanks for reaching out.

We are indeed aware of the updates to EIP 1193, because I wrote the PR that deprecated `send` in favor of a new method, `request`.

The updates to EIP 1193 followed discussions by the original authors of that EIP, convenience library authors (`ethers`, `web3`), dapp developers, and other community members. They were designed to make the new API _less breaking_ for dapps and convenience libraries, and _easier to implement_ for wallets.

Our documentation and the above Medium articles will be updated to reflect the new changes sometime tomorrow, April 6, Pacific Time.

The ship date for the new API is still TBD, and this issue will be updated accordingly.

**Comment by crazyrabbitLTC** [comment]
Hello all, is there perhaps a timeline we can refer to regarding the new API? Looking to understand when we're going to need to have resources to build out the new support.

**Comment by alextreib** [comment]
I agree with @crazyrabbitLTC, we would appreciate a timeline, since it said early 2020. Now, we are almost in Mid 2020.

**Comment by bhayward93** [comment]
Very much in the same boat as @crazyrabbitLTC and @alextreib here. Further details highlighting a predicted timeline would be much appreciated, even if not totally accurate, since we can then roughly allocate resources.

**Comment by rekmarks** [comment]
We are finally able to give an update about this!

### Action Items

- If you rely on the `window.web3` object injected by MetaMask, you are going to have to make changes after August 1 (2020-08-01), or your dapp will break.
  - We recommend migrating to using `window.ethereum` or `ethers`, ASAP.
  - More details about this in an upcoming announcement.
- If you rely on the `window.ethereum` object injected by MetaMask, you may have to make changes in the next few months. Read on for details.
- If you only rely on your own version of `ethers`, you don't have to do anything except keep it up to date. `web3` should mostly be fine too. If you use a convenience library, we recommend `ethers`.

We will ship a _non-breaking_ update that includes [the new Provider API](https://eips.ethereum.org/EIPS/eip-1193) in the next few weeks, and will make an announcement when the time comes. Read on for details.

### Changes in Detail

- As of right now, the Provider API (`window.ethereum`) will _not_ undergo any breaking changes, _except_ for some experimental methods on `ethereum._metamask`.
- The following breaking changes will occur in the next few months:
  - We will stop reloading the page on network change
  - We will stop emitting `chainIdChanged`, and instead emit `chainChanged`
  - We will remove the following experimental methods:
    - `ethereum._metamask.isEnabled`
    - `ethereum._metamask.isApproved`
- The following breaking changes will occur at some point after August 1, 2020:
  - We will stop injecting `web3` into web pages

We will ship the new Provider API in the next few weeks, without any breaking changes, and will make an announcement when the time comes.

### Background

After our original announcement, there was a movement to make further changes to the standard defining the Provider API, [EIP 1193](https://eips.ethereum.org/EIPS/eip-1193). We were deeply involved in writing those updates, and helped create a version that is non-breaking. This means that, if you rely on `ethereum.send` or `ethereum.sendAsync` today, they will continue to work. Nevertheless, we think the new API is pretty sweet, and offers some cool features, and we recommend that you check it out [here](https://eips.ethereum.org/EIPS/eip-1193).

**Comment by aksdevac** [comment]
@rekmarks  when this new api will be released? as new dapps also need to use old API currently. so if is there any timeline it would be helpful.

**Comment by Kharabet** [comment]
@rekmarks I have updated MetaMask extension to v. 7.7.9 and still get old-signature method `ethereum.send(payload, callback)` at some pages instead of new `ethereum.send(method: string, params: Array<any>)`. What can cause different versions MetaMask APIs injecting?

**Comment by rekmarks** [comment]
@aksdevac, as early as next week, but probably the week of May 18, 2020.

**Comment by rekmarks** [comment]
For everyone's knowledge, the issue raised by @Kharabet was due to `web3@1.x` overwriting `ethereum.send`. This will _not_ happen with `ethereum.request`, which we will prevent overwriting via a Proxy.

**Comment by samajammin** [comment]
Thanks for the detailed update @rekmarks!

I'm curious - why the decision to "stop reloading the page on network change"? This is something convenience libraries (like `ethers`) will not be able to easily handle. I agree with Richard Moore's [comment here](https://github.com/ethers-io/ethers.js/issues/589#issuecomment-630725294) on why MetaMask _should_ continue to reload the page on network change.

**Comment by adrianmcli** [comment]
@samajammin It's more hassle for the developer but it provides better end-user UX. It allows the user to keep their app state and not lose context. Regarding Ethers.js (it's essentially the same issue w/ Web3.js), you simply acquire a new provider once you get the event.

Here's some recent code I worked on just the other day to do this w/ React hooks:

```js
// re-register MetaMask provider whenever network changes
useEffect(() => {
  window.ethereum.on("chainChanged", (network) => {
    const provider = new ethers.providers.Web3Provider(window.ethereum);
    setProvider(provider)
  });
}, [provider]);
```

**Comment by balasan** [comment]
> For everyone's knowledge, the issue raised by @Kharabet was due to web3@1.x overwriting ethereum.send. This will not happen with ethereum.request, which will throw an error on overwrite attempts via a Proxy.

@rekmarks is there currently any way to prevent this from happening? Will switching to `request` be the only way to avoid this going forward?

**Comment by rekmarks** [comment]
@samajammin, thanks! Regarding reloading on network change, I think @ricmoo makes a convincing argument that it's often a good idea to do so, but we (MetaMask) don't believe that we should make that choice for others.

Regarding changing networks and the average user, it is not true that only dapp devs use test networks, as evidenced by e.g. Reddit's rewards. We're definitely open to the idea of a "dev mode" in MetaMask in the future, but believe that our users are already protected because:
- Changing the network is an unambiguous action in the MetaMask UI
- The selected network is prominently displayed in the MetaMask UI whenever it is relevant

> This [not reloading on network change] is something convenience libraries (like `ethers`) will not be able to easily handle.

With the new Provider API, library and dapp developers can implement this with a single line:
```
ethereum.on('chainChanged', () => window.location.reload())
```

I fully support anyone that wants to continue reloading pages on network changes, we're just not going to force that behavior on you by default. We'll continue to recommend `ethers` regardless!

**Comment by rekmarks** [comment]
@balasan, AFAIK, there's no way to prevent `web3@1.x` from overwriting `ethereum.send` at the moment, other than not using `web3@1.x`. The version injected by MetaMask — _which, I'll repeat, everyone should stop using immediately because we are going to stop injecting it later this year_ — does not overwrite `ethereum.send`.

`ethereum.request` is absolutely the recommended method going forward. Our provider will throw an error if an overwrite of `ethereum.request` is attempted, and deletions will be ignored.

**Comment by alextreib** [comment]
Please keep the web3 API alive as long as the ethereum isn't completely defined and has bugs.

**Comment by rekmarks** [comment]
@alextreib we won’t start the 3-month `web3` deprecation period until EIP-1193 has been finalized. At the moment, it looks like this will happen on 2020-06-04. Either way, it will be announced here.

Regardless of 1193, you should stop using our injected version of `web3` today. We recommend `ethers`.

**Comment by ricmoo** [comment]
Sorry... Late to the game. I'm writing the docs on the best way to restore the behaviour of reloading the page in the "best practices" section of the docs right now. :)

So chiming in with my two cents, feel free to disregard, but this exact thing is leading to other issues, which are terribly difficult to solve (possibly undecidable) in a generic way.

I think the default of reloading still makes sense. It is fine to allow an option to override this behaviour for advanced cases where that makes sense, especially in developer tools (e.g. remix would likely want to handle this).

I would also say that unless the developer is explicitly listening to the "chainChanged" event, the behaviour should not be able to be changed from this default refresh. If a developer is not explicitly listening for the network to change, they are likely not equipped to handle it. If they really know what they are doing, they can register with an empty callback.

I guess I wonder what the value for the default behaviour to not refresh is? As long as there is a way for a developer to announce they are advanced and wish to tackle the nuances, that should be fine. Refreshing ensures any app (ish) works in general across networks, even for inexperience (and more importantly non-veteran) developers. Not refreshing can lead to lost funds, leaking data and inconsistent non-atomic operations.

I'll just give a few examples of small things that become very hard to handle if a dapp developer doesn't fully understand the consequences and minutia of handling a network change (at which time they are responsible to unregister all listeners, purge all caches, and re-register for events, handle in-flight activities).

1. Listening on an event filter; when the network changes, all callbacks may now need to be de-registered, and re-registered on the new address. As this often happens within a closure, the closure must be re-evaluated. That closure must also have either exposed some way to de-register or itself detect network changes and pull itself out, but manage that it pulls itself out before the new closure registers.
2. Listening for block events; when the network changes the current block number could either go back far in the past, jump far into the future or stay similar, each of these cases need to be handled possibly differently, deepening on the app.
3. Listening for Fork events; a once mined transaction now becomes unmined, possibly beyond what would normally be considered safe number of confirmations, which means it may go undetected. The transaction may also have different meaning on this new network, or transactions may still have occurred, but in a different order.

Basically, Turing Complete is hard to deal with. So, the default should be as safe as possible, for the widest audience of developers, and people who want the additional exercise, may do so, but opt-in for it.

That's my protest for now. :)

**Comment by crazyrabbitLTC** [comment]
I would also like to throw my voice in with regards to refreshing the page. As long as people can switch providers, which often leads to unpredictable outcomes, I also think we should refresh the page. It's so very hard to keep in sync with a network that is changing, non-deterministic, with nodes that might return conflicting data that allowing the user to change networks suddenly without reloading the page just adds to the weight of developers trying to catch edge cases. Sure it's not a nice experience, but if you're changing the network, you're doing something pretty dramatic. It's like connecting to a whole different internet. 

So, just another voice.

**Comment by rekmarks** [comment]
Yeah, we discussed this pretty extensively internally, and are figuring out what to do. I don't want to make any premature announcements, but unless you have _an incontrovertible need to handle network changes without refreshing the page_, our official recommendation is to do:
```javascript
ethereum.on('chainChanged', () => window.location.reload())
```

**Comment by alcuadrado** [comment]
> but unless you have _an extremely good reason to handle network changes without refreshing the page_, our official recommendation is to do:
> 
> ```js
> ethereum.on('chainChanged', () => window.location.reload())
> ```

This makes me think that this API is designed for the least frequent use-case by default. Won't this lead to the majority of apps accidentally breaking? Not talking just of existing ones, but also new ones.

I think giving devs the possibility of not having their pages reloaded on network changes is **really** important. I'm glad you are adding that. I just worry that it may be yet another hard-to-discover gotcha of the Ethereum platform.

> I would also say that unless the developer is explicitly listening to the "chainChanged" event, the behaviour should not be able to be changed from this default refresh.

This idea is really interesting.

**Comment by rekmarks** [comment]
Alright, here's the complete context:

We were convinced by @ricmoo's argument above, that the _default_ behavior should be that the page is reloaded. I also agree with the points raised by @crazyrabbitLTC and @alcuadrado. We also want consumers to be able to override the reload, if they wish.

Now, the problem is that we _don't_ reload by default under most circumstances, at the moment. We only reload by default if our injected `window.web3` object has been accessed. (Don't ask why. 😑) What we want to do is reload if our `window.ethereum` object, the provider, has been accessed.

That being the case, enabling reload by default would probably be breaking for existing dapps. Probably not catastrophically breaking, but breaking nonetheless. That being the case, we need to go through the entire deprecation process. We'll most likely add it to the first round of changes noted previously in this thread, 6 weeks after the launch of MetaMask v8.

Regarding how to toggle off the default, the majority opinion on the team is to have an explicit toggle for the behavior, rather than an implicit one, e.g. by adding a `chainChanged` listener.

**Comment by ricmoo** [comment]
Random idea... Not sure if it's good or not, but just thinking out loud. Rather than having functions on the ethereum object, use something like: `ethereum.request("com.metamask_stopReloadOnNetworkChange", [])`?

I would also like to start (or help start?) an EIP on vendor extensions, like this. I also plan to update [EIP-634](https://eips.ethereum.org/EIPS/eip-634) to use a similar reverse dot notation, so there is no central registry required, while preserving hierarchal vendor types.

**Comment by dkent600** [comment]
I am at a loss to find docs on `ethereum.request`.  Can someone please refer me?  Thanks!

**Comment by dkent600** [comment]
> I am at a loss to find docs on `ethereum.request`. Can someone please refer me? Thanks!

Found some:  https://eips.ethereum.org/EIPS/eip-1193#request

**Comment by snario** [comment]
_(slightly adjacent topic)_ I just want to throw this in here in case the right person sees this - are there any community adopted TypeScript typings around the `ethereum` object that exist today and would be expected to be upgraded as this issue and associated EIP develops?

**Comment by jacobcantele** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [web3.js deprecation comms](https://trello.com/c/OthdD2LZ/109-web3js-deprecation-comms)

**Comment by jacobcantele** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [web3.js deprecation comms](https://trello.com/c/OthdD2LZ/109-web3js-deprecation-comms)

**Comment by snario** [comment]
![image](https://user-images.githubusercontent.com/1933029/86786210-bcb30200-c031-11ea-9327-ba31b82e4d62.png)

**Comment by karega** [comment]
What should developers do that were using the web3.js 0.2 version that was being injected?  I was using this to sign transactions directly using the PK.

**Comment by DeltaBalances** [comment]
Maybe this status needs an update.  
> The New API is: Not Live  

Looks like it is partially live. I can currently use `ethereum.request({ method: 'eth_requestAccounts' })` on Firefox but not yet on Chrome.

The [MetaMask docs](https://docs.metamask.io/guide/ethereum-provider.html#using-the-provider) show a rollout warning.
![rollout](https://user-images.githubusercontent.com/30391324/87370056-44dd5e00-c582-11ea-8539-a5eb75cf2d58.png)

**Comment by rekmarks** [comment]
The new API is _live_ on Chrome and Firefox. Please see [the original issue comment](https://github.com/MetaMask/metamask-extension/issues/8077#issue-568472880) for details.

It's also MetaMask's 4th birthday (🦊), and Bastille Day (🇫🇷). Happy birthday to us, and _bonne fête nationale_ to the French!

**Comment by rekmarks** [comment]
Reposting latest announcement for great email notification convenience:

### The New API is: **_Live_**

The new Provider API is live in MetaMask 8.0.4 on Chrome and Firefox as of 2020-07-14, which happens to be MetaMask's 4th birthday 🎉 .

Please consult [our documentation](https://docs.metamask.io/guide/ethereum-provider.html) to learn how to use the API.

The new API should be _non-breaking_ for all dapps. If your dapp is broken, it is an all likelihood a bug. Please [file a bug report](https://github.com/MetaMask/metamask-extension/issues/new?template=bug-report.md) if so.

### Upcoming Breaking Changes

We are still going to make _some_ breaking changes to our provider. These include:

- Stop 0-padding `eth_chainId` return values
- Stop emitting `chainIdChanged`, and instead emit `chainChanged`
- Remove the following experimental methods:
  - `ethereum._metamask.isEnabled`
  - `ethereum._metamask.isApproved`

We are also still going to remove our injected `window.web3`.

Please see [the relevant section of our documentation](https://docs.metamask.io/guide/ethereum-provider.html#upcoming-breaking-changes) for details. _You can migrate to protect yourself from these changes today._ Please see our [migration guide](https://docs.metamask.io/guide/provider-migration.html) for details. We will announce dates soon, but there's no reason to wait. [Migrate today!](https://docs.metamask.io/guide/provider-migration.html)

**Comment by rekmarks** [comment]
Since MetaMask v8 launched, we noticed that the `ethereum.publicConfigStore` object is being used by some consumers. We've also seen it accessed as `web3.currentProvider.publicConfigStore`.

Despite its name, the `publicConfigStore` object was never intended for public consumption, and _it is scheduled for removal_ along with the other `window.ethereum` API changes mentioned in [the original issue comment](https://github.com/MetaMask/metamask-extension/issues/8077#issue-568472880) (which has been updated with this removal).

Everything you can do with the `publicConfigStore` can be done at least as easily via other properties / events / methods / public state etc. If you are using the `publicConfigStore`, please review why, and replace it with the corresponding APIs in [our documentation](https://docs.metamask.io/guide/ethereum-provider.html#upcoming-breaking-changes).

**Comment by klesun** [comment]
Our app stopped working today with `8.07`. I thought compatibility breaking changes were planned on 30 of August?

**Comment by rekmarks** [comment]
@klesun I was unable to reproduce your issue locally using 8.0.7. We have not yet made any breaking changes to the provider API.

I recommend trying `ethereum.request({ method: 'eth_requestAccounts' })` instead of `ethereum.send` or `ethereum.enable`. Please create a separate issue if the problem persists!

**Comment by rekmarks** [comment]
@karega, to answer your question about what to do if you're using the injected `window.web3`, please see our [migration guide](https://docs.metamask.io/guide/provider-migration.html). I apologize for the delayed response!

**Comment by klesun** [comment]
Though `8.0.6` did work with our dapp ok today, so I take it that some breaking changes _were_ introduced in `8.0.7`. I'll update my comment when I find the cause.

**Comment by rekmarks** [comment]
@klesun That's strange, because we didn't ship any changes to the provider in 8.0.7. See: https://github.com/MetaMask/metamask-extension/compare/v8.0.6...v8.0.7

We'd love to understand what the problem is, but please create a new issue instead of updating your previous comment.

**Comment by rekmarks** [comment]
### Breaking Changes Shipping on November 16, 2020

The breaking changes will ship on or shortly after **November 16, 2020**. Please see the [original issue comment](https://github.com/MetaMask/metamask-extension/issues/8077#issue-568472880) for details.

As announced previously, the new API is live, as are tools for moving off our injected `window.web3`. There's no reason to wait, [migrate today!](https://docs.metamask.io/guide/provider-migration.html)

**Comment by rekmarks** [comment]
If you want to try out your dapp with a version of MetaMask that does not inject `window.web3`, you can grab the latest build from #9156 (posted automatically by [@metamaskbot](https://github.com/metamaskbot)) and follow the [instructions in our README](https://github.com/MetaMask/metamask-extension/#other-docs) for loading a custom build of MetaMask.

Remember to only load custom MetaMask builds from trusted sources, like our GitHub!

**Comment by dominicletz** [comment]
Regarding eth_chainId the check `} else if (chainId !== endpointChainId) {` in [network-form.component.js](https://github.com/MetaMask/metamask-extension/blob/af018a2d746d57080f4d06acda93dfa97dc56ab3/ui/app/pages/settings/networks-tab/network-form/network-form.component.js#L275) seems is using strings and thus is very brittle. Any thought on converting the hex to numbers before the comparison, so that numbers are compared and not formatted strings?

**Comment by rekmarks** [comment]
@dominicletz I sympathize with your point, but it is brittle by design. We do not want to be in a position where we or consumers of our provider API have to parse `chainId` values. I'll refer to [EIP-695](https://eips.ethereum.org/EIPS/eip-695) - which specifies the `eth_chainId` RPC method - and the [Ethereum JSON-RPC API](https://eth.wiki/json-rpc/API), which specifies data types for Ethereum JSON-RPC methods.

You'll note that, per EIP-695, the `chainId` value is a `QUANTITY`, which the Ethereum JSON-RPC API states must not have leading zeros (see the "HEX value encoding" heading in the second link above).

**Comment by dominicletz** [comment]
> @dominicletz I sympathize with your point, but it is brittle by design. We do not want to be in a position where we or consumers of our provider API have to parse `chainId` values. I'll refer to [EIP-695](https://eips.ethereum.org/EIPS/eip-695) - which specifies the `eth_chainId` RPC method - and the [Ethereum JSON-RPC API](https://eth.wiki/json-rpc/API), which specifies data types for Ethereum JSON-RPC methods.
> 
> You'll note that, per EIP-695, the `chainId` value is a `QUANTITY`, which the Ethereum JSON-RPC API states must not have leading zeros (see the "HEX value encoding" heading in the second link above).

Uh! Thanks for that note, wasn't even aware.

**Comment by rekmarks** [comment]
You're very welcome, thanks for understanding!

**Comment by farhanJR** [comment]
After doing all the migration updates mentioned and trying to send the transaction, metamask pops-up but when I click confirm, the website is not able to get the transaction hash from MM. 

I used the same code as mentioned on the MM Docs, i.e, 

`const txHash = await ethereum.request({
  method: 'eth_sendTransaction',
  params: [transactionParameters],
});`

also, this random error is being poping up on Dev Console:

`ERROR Error: Uncaught (in promise): TypeError: this._processRequest(...).catch(...).finally is not a function
TypeError: this._processRequest(...).catch(...).finally is not a function
    at e._handle (inpage.js:15)
    at e.handle (inpage.js:15)
    at t.exports._rpcRequest (inpage.js:1)
    at inpage.js:1
    at new ZoneAwarePromise (zone.js:875)
    at t.exports.request (inpage.js:1)
    at Web3Service.<anonymous> (web3.service.ts:1452)`

at line 1452, on web3service, we have: 

`window['ethereum'].request({
          method: 'eth_sendTransaction',
          params: [
            {
              to: <Address>,
              from: <currentWallet>,
              value: "0x0",
              data: <encodedData>
            },
          ],
        }).then((txHash) => console.log(txHash))
        .catch((error) => console.error);
`

Any help/input would be amazing

**Comment by farhanJR** [comment]
@rekmarks 

On this migration, For interacting with Contract on the Angular website we used to do it like this:

`this.web3.eth.contract(<abi>).at(<address>)` & then use the contract-instance object to interact, but with the MM migration, how do we go about doing the contract interaction??

Also, I updated my implementation to use `ethereum.request({method: 'eth_sendTransaction',params: []})`, but, when I use it now, it doesn't work and gives me the below error:

`e.abi.filter is not a function at c (inpage.js:15)`

Also, I have multiple places where contract interaction occurs, any help or guidance would be amazing :) 

Just want to understand this contract interaction bit after MM migration, as Guide doesn't talk about it.

**Comment by drumenbassik** [comment]
> ### Breaking Changes are Coming _November 16, 2020_
> We are removing `window.web3` and making some breaking changes to `window.ethereum`. These changes apply to all version of MetaMask (extension and mobile). See below for details.
> 
> ### New Provider API
> We shipped a new, _non-breaking_ Provider (`window.ethereum`) API in MetaMask 8.0.4 on July 14, 2020, which happened to be MetaMask's 4th birthday 🎊 🦊 🎊. This API is also available through MetaMask mobile.
> 
> Please consult [our documentation](https://docs.metamask.io/guide/ethereum-provider.html) to learn how to use the API.
> 
> ### Upcoming Breaking Changes
> These changes will ship on or after **November 16, 2020**. The version number of MetaMask that includes the breaking changes will be announced by this date.
> 
> #### `window.web3` Removal
> We are going to remove our injected `window.web3`. See #7163 for why, and our [migration guide](https://docs.metamask.io/guide/provider-migration.html#replacing-window-web3) for replacement strategies.
> 
> See #9156 for the latest build without `window.web3`, automatically posted by [@metamaskbot](https://github.com/metamaskbot).
> 
> #### `window.ethereum` API Changes
> We are also going to make _some_ breaking changes to our provider. These include:
> 
> * Stop 0-padding `eth_chainId` return values
> * Stop emitting `chainIdChanged`, and instead emit `chainChanged`
> * Remove `ethereum.publicConfigStore`
> * Remove the following experimental methods:
>   
>   * `ethereum._metamask.isEnabled`
>   * `ethereum._metamask.isApproved`
> 
> #### What to Do Next
> Please see [the relevant section of our documentation](https://docs.metamask.io/guide/ethereum-provider.html#upcoming-breaking-changes) for details. _You can migrate to protect yourself from these changes today._ Please see our [migration guide](https://docs.metamask.io/guide/provider-migration.html) for details. There's no reason to wait, [migrate today!](https://docs.metamask.io/guide/provider-migration.html)

**Comment by rekmarks** [comment]
It is now November 16, and we will stop injecting `window.web3` and ship the breaking changes to our `window.ethereum API` in an upcoming release.

Due to app store review processes, we can’t say exactly when the changes will go live, but you should no longer rely on any of the affected features, as they may break at any time. Both the MetaMask browser extension and mobile app will be affected.

We will make an announcement here when the changes have started to roll out, along with the version number that includes the changes.

**Comment by octofi** [comment]
Thanks for your heads up.

**Comment by rekmarks** [comment]
You can now grab test builds here: #10015

That build includes all breaking changes. We will update it with the latest changes until we ship them to production. See the [extension readme](https://github.com/MetaMask/metamask-extension/#other-docs) for how to install a custom build, and  consult [the migration guide](https://docs.metamask.io/guide/provider-migration.html) if you discover any issues with your dapp.

**Remember to only download custom builds posted by @metamaskbot on our GitHub.**

**Comment by rekmarks** [comment]
The breaking changes will ship after the US holiday season, in January of the New Year. See y'all in 2021 🎆.

**Comment by rekmarks** [comment]
The breaking changes will be part of version `9.0.0` of the extension, likely out the next week (starting January 11). You can now grab test builds from the release candidate: #10054

**Comment by rekmarks** [comment]
Firefox will receive the breaking changes today. Remaining browsers over the next few days, possible at the beginning of next week (January 18), depending on the extension store review processes.

**Comment by rekmarks** [comment]
A `9.0.1` patch release will ship to Firefox ASAP. Chrome will receive `9.0.1` directly, pending the Chrome store review process.

**Comment by rekmarks** [comment]
`9.0.1` is currently at 100% Firefox and 20% Chrome. As reported in #10216, `web3@1.x` is currently broken for users that have received the `9.0.1` update. We're publishing a `9.0.2` patch in order to help restore functionality for users of `web3@1.x`, and it will roll out over the next few days.

The `web3.js` team (i.e. ChainSafe) are shipping version `web3@1.3.2`, which will be fully compatible with all `9.0.x` versions of the MetaMask extension. If your website depends on `web3@1.x`, we encourage you to update to `1.3.2` as soon as it's ready. You can follow their releases here: https://github.com/ChainSafe/web3.js/releases

**Comment by rekmarks** [comment]
[`web3@1.3.2`](https://github.com/ChainSafe/web3.js/releases/tag/v1.3.2) has shipped. We recommend upgrading ASAP if you use `web3@1.x`.

`9.0.2` of the MetaMask extension is live on Firefox as of ~24 hours ago, and will go live on Chrome as soon as the Chrome store review process is completed.

In other news, [MetaMask Mobile](https://github.com/MetaMask/metamask-mobile) is also in the process of shipping the `window.web3` removal and breaking provider API changes. Once `9.0.2` has landed on Chrome, we will keep this issue open for one more week, after which we'll consider the rollout complete.

**Comment by rekmarks** [comment]
**The breaking changes have shipped on all platforms.**

(With the possible exception of the Edge store, but that will happen soon enough, and its user base is very small.)

For the benefit of any latecomers, we'll keep this issue open for another week. Thank you for tuning in, and for your patience during the deprecation period. Cheers!

**Comment by arbin304** [comment]
Is there any documentation on how window.ethereum api is created ? I couldnt find it anywhere in the code

**Comment by rekmarks** [comment]
@arbin304 you can find our provider implementations in this repository: https://github.com/MetaMask/providers

For the API reference, we recommend our docs: https://docs.metamask.io/guide/ethereum-provider.html
